### PR TITLE
 Clava modernisation and API conversion to TypeScript

### LINKED ISSUE #48: Make the Clava GUI sidebar available in other IDEs.
Explore making an extension for vscode or other IDEs that provides the same information as the sidebar in the Clava editor.

### PR DESCRIPTION
**Pull request by lm-sousa**
**[WARNING] Merging of this PR is dependant on the prior merge of specs-feup/lara-framework#41**

Some of the Clava API files have been translated into TypeScript.

These changes include:

- Creation of the Lara-JS project.
- Moving APIs to this new project.
- Ability to export APIs in JS form from Clava-JS to Clava Classic.
- Class member information in IDEs regarding Joinpoints is now available through the use of wrapper classes.
- APIs are now properly recognized by IDEs and its features can now aid in development.
- Scripts are provided to automatically generate the wrapper classes from the JSON specification created by ClavaWeaverGenerator.
- APIs created by JulesHervault during his internship were moved from Lara to Clava as its contents were dependant on Clava specific features.
- The `MathExtra` API has been moved from Lara to Clava as it is Clava specific.
- Generation of a website containing documentation of APIs present in the Clava-JS project has been implemented.
- Closes #48 as it is now obsolete.
- In the Language Specification, when user-defined enums types are used in joinpoints methods, their implementation uses the type String instead of the Java enum type. This is to allow for interoperability with JavaScript.

## Not addressed in this PR

- Not all APIs have been translated into TypeScript.
- Documentation of the new development workflow.
- The CI flow for automatic generation and publishing of the API documentation website.

### Joinpoint Specification

Modifications were made to the Joinpoint Specification of Clava to make in line with the new conventions and keep compatibility with the new version of Lara Framework (specs-feup/lara-framework#41)

<details><summary>Details</summary>
<p>

- Joinpoint
  - `ancestor` --> `getAncestor`
  - `astAncestor` --> `getAstAncestor`
  - `chainAncestor` --> `getChainAncestor`
  - `descendants` --> `getDescendants`
  - `descendantsAndSelf` --> `getDescendantsAndSelf`
  - `laraDescendants` --> (deleted)
  - `astParent` --> (deleted)
  - `hasAstParent` --> (deleted)
  - `child` --> `getChild`
  - `astChild` --> `getAstChild`
  - `firstJp` --> `getFirstJp`
  - `value` --> (deleted)
  - `userField` --> (deleted)
  - `joinpointType` --> (deleted)
  - `javaValue` --> (deleted)
  - `javaFieldType` --> `getJavaFieldType`
  - `keyType` --> `getKeyType`

- Call
  - `arg` --> `getArg`

- File
  - `destinationFilepath` --> `getDestinationFilepath`

- Function
  - `declaration` --> `getDeclaration`

- Loop
  - `changeKind` --> (deleted)

- Omp
  - `reduction` --> `getReduction`

- Pragma
  - `targetNodes` --> `getTargetNodes`

- Scope
  - `numStatements` --> `getNumStatements`

- Type
  - `bitWidth` (variant that accepted a parameter) --> (deleted)
</p>
</details> 

### Lara Common Language

The changes introduced by this branch break compatibility with the 'Lara Common Language' experimental project.
Compatibility could be re-established by having 'Lara Common Language' register its own wrappers with Lara for use with the `wrapJoinPoint` and `unwrapJoinPoint` methods. Although, there is a case to be made for dropping support entirely and spinning off the 'Lara Common Language' project. It may make more sense than to include an experimental feature in mainline Clava.

### Others

- Usage of `$jp.userFields = {...}` has been discontinued. Refer to `joinpoint.setUserFields`
- Also, in-place updating of joinpoint user fields is **no longer possible** due to the use of wrappers (e.g. `$jp.userField("myArray").push("foo")`).
- `JoinPoints.toJoinPoint(node)` has been moved to `ClavaJoinPoints.toJoinPoint(node)`
- The `ClavaNodes` API alias of `ClavaJoinPoints` has been removed as it was not being used and had a deceptive name.

### CODE REVIEW

File: ClavaWeaver/resources/clava/test/api/TimerTest.lara [hunk 1]
```diff
@@ -24,13 +25,7 @@ aspectdef TimerTest
...
 	apply
-		var bar2TimeVar = timer.time($call);		
-		var logger = new Logger();
-		logger.text("I want to print the value of the elapsed time (")
-			.double(bar2TimeVar)
-			.text("), which is in the unit '" + timer.getUnit().getUnitsString() + "' and put other stuff after it")
-			.ln()
-			.log(timer.getAfterJp());
+		var bar2TimeVar = timer.time($call);
 	end
```
Comment (joaobispo) [inline ClavaWeaver/resources/clava/test/api/TimerTest.lara:29]: Why did this disappeared?

File: ClavaWeaver/src/pt/up/fe/specs/clava/weaver/CxxWeaver.java [hunk 2]
```diff
@@ -2099,5 +2098,13 @@ public AstMethods getAstMethods() {
...
                 node -> ClavaCommonLanguage.getJoinPointName(node), node -> node.getScopeChildren());
     }
+    
+    @Override
+    public List<LaraResourceProvider> getNpmResources() {
+        return Stream.concat(
+                super.getNpmResources().stream(),
+                Arrays.asList(ClavaApiJsResource.values()).stream())
+            .toList();
+    }
```
Comment (joaobispo) [inline ClavaWeaver/src/pt/up/fe/specs/clava/weaver/CxxWeaver.java:2109]: Verify if in future PRs (or staging) you still need to explicitly call super.getNpmResources()

**Review by joaobispo — APPROVED** [review]

File: Clava-JS/src-api/clava/ClavaJoinPoints.ts [hunk 3]
```diff
@@ -0,0 +1,857 @@
...
+      $type = ClavaJoinPoints.typeLiteral($type);
+    }
+
+    return wrapJoinPoint(
+      ClavaJavaTypes.AstFactory.ternaryOp(
+        unwrapJoinPoint($cond),
+        unwrapJoinPoint($trueExpr),
+        unwrapJoinPoint($falseExpr),
+        unwrapJoinPoint($type)
+      )
```
[the remark this answers is not in the document]
Comment (lm-sousa) [reply Clava-JS/src-api/clava/ClavaJoinPoints.ts:565]: Perhaps but I doubt we could maintain most helpers from the compiler on these calls.

File: ClavaWeaver/resources/clava/test/api/TimerTest.lara (hunk 1 above)
Comment (lm-sousa) [reply ClavaWeaver/resources/clava/test/api/TimerTest.lara:29]: Re-added in 4de9ebb50

File: Clava-JS/src-api/clava/ClavaCode.ts [hunk 4]
```diff
@@ -77,29 +86,28 @@ class ClavaCode {
...
+    for (const child of $file.children) {
+      if (child instanceof FunctionJp && child.storageClass === StorageClass.STATIC) {
+        const newName = child.name + "_static_rename";
         child.name = newName;
         changedCode = true;
       }
 
-      if (child.astName === "DeclStmt" && child.storageClass === "static") {
-        println(child.code);
+      if (child instanceof Vardecl && child.storageClass === StorageClass.STATIC) {
```
[the remark this answers is not in the document]
Comment (lm-sousa) [reply Clava-JS/src-api/clava/ClavaCode.ts:109]: The `storageClass` attribute is only available in the `FunctionJp` and `VarDecl` joinpoints and the classes that inherit from them. The previous `if` statement checks if the joinpoint is a `FunctionJp`. Given that, the option that remains is to check if it is a `VarDecl`.

I assume this bug has been present in the code for years and this shift to Typescript has allowed us to finally catch it.

[the remark this answers is not in the document]
Comment (lm-sousa) [reply Clava-JS/src-api/clava/ClavaCode.ts:109]: For the record: `DeclStmt` does not extend either `Vardecl` or `FunctionJp`.

**Pull request merged**